snap/validate.go

```diff
@@ -263,6 +270,193 @@ func validateAppSocket(socket *SocketInfo) error {
+				// add only if the other does not list this one
+				// as Before
+				graphAppMap[after].Before = append(graphAppMap[after].Before,
+					graph[i].Name)
+			}
+		}
+		graph[i].After = nil
+	}
+
+	// count of all edges in a graph
+	edges := 0
+
+	for i := range graph {
+		// 'before' count as incoming edge of other:
+		//    app --(before)--> other
+		for _, other := range graph[i].Before {
+			indegrees[other] += 1
+			edges += 1
+		}
+	}
+
+	// queue with nodes that have no incoming edges
+	queue := make([]*auxApp, 0, len(graph))
+
+	// fill it
+	for i := 0; i < len(graph); i++ {
+		app := graph[i]
+		if indegrees[app.Name] == 0 {
+			queue = append(queue, app)
+		}
+	}
+
+	// Kahn:
+	// - queue: nodes without incoming edges
+	// - sorted: sorted nodes
+	//
+	// Nodes without incoming edges are 'top' nodes and are appended to
+	// 'sorted'. On each iteration, take the next 'top' node, and decrease
+	// each adjecent node's indegree (count of incoming edges). Once that
+	// adjecent node's indegree is 0 (no incoming edges) is can become a
+	// 'top' node and is put on the queue. While doing so, decrease the edge
+	// count.
+	for len(queue) > 0 {
+		u := queue[0]
+		queue = queue[1:]
+		sorted = append(sorted, u.App)
+		for _, before := range u.Before {
+			indegrees[before] -= 1
+			edges -= 1
+			if indegrees[before] == 0 {
+				queue = append(queue, graphAppMap[before])
+			}
+		}
+	}
+
+	// if our graph is a DAG, then we were able to account for all incoming
+	// edges of each node
+	if edges != 0 {
+		// some edges still left, thus not a DAG, raise an error
+		unsatisifed := bytes.Buffer{}
+		for name, v := range indegrees {
+			if v > 0 {
+				if unsatisifed.Len() > 0 {
+					unsatisifed.WriteString(", ")
+				}
+				unsatisifed.WriteString(name)
+			}
+		}
+		return sorted, fmt.Errorf("dependency cycle detected for apps %q",
+			unsatisifed.String())
+	}
+	return sorted, nil
+}
+
+func validateAppsOrdering(snap *Info) error {
+	apps := make([]*AppInfo, 0, len(snap.Apps))
+	for _, app := range snap.Apps {
+		apps = append(apps, app)
+	}
+	_, err := orderBeforeAfter(apps)
+	if err != nil {
+		return fmt.Errorf("cannot validate app startup ordering: %v",
+			err.Error())
+	}
+	return nil
+}
+
+func validateAppStartupOrdering(app *AppInfo, order appStartupOrder) error {
+	ordering := appStartupOrdering(app, order)
+
+	// we must be a service to request ordering
+	if len(ordering) > 0 && !app.IsService() {
+		return fmt.Errorf("cannot validate app %q: requires startup ordering, but is not a service",
+			app.Name)
+	}
+
+	for _, dep := range ordering {
+		// dependency is not defined
+		other, ok := app.Snap.Apps[dep]
+		if !ok {
+			return fmt.Errorf("cannot validate app %q startup ordering: needs to be started %s %q, but %q is not defined",
+				app.Name, order, dep, dep)
+		}
+
+		if !other.IsService() {
+			return fmt.Errorf("cannot validate app %q startup ordering: dependent app %q is not a service",
```

Comment:
This is also a little bit verbose IMHO, "cannot validate app %q: %q is not a service" would work too. But again, let's check with Gustavo.